bindings/influx/influx.go

```diff
@@ -0,0 +1,107 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package influx
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/dapr/pkg/logger"
+	influxdb2 "github.com/influxdata/influxdb-client-go"
+	"github.com/influxdata/influxdb-client-go/api"
+)
+
+// Influx allows writing to InfluxDB
+type Influx struct {
+	metadata *influxMetadata
+	client   influxdb2.Client
+	writeAPI api.WriteAPIBlocking
+	logger   logger.Logger
+}
+
+type influxMetadata struct {
+	URL    string `json:"url"`
+	Token  string `json:"token"`
+	Org    string `json:"org"`
+	Bucket string `json:"bucket"`
+}
+
+// NewInflux returns a new kafka binding instance
+func NewInflux(logger logger.Logger) *Influx {
+	return &Influx{logger: logger}
+}
+
+// Init does metadata parsing and connection establishment
+func (i *Influx) Init(metadata bindings.Metadata) error {
+	influxMeta, err := i.getInfluxMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	i.metadata = influxMeta
+	if i.metadata.URL == "" {
+		return errors.New("Influx Error: URL required")
+	}
+
+	if i.metadata.Token == "" {
+		return errors.New("Influx Error: Token required")
+	}
+
+	if i.metadata.Org == "" {
+		return errors.New("Influx Error: Org required")
+	}
+
+	if i.metadata.Bucket == "" {
+		return errors.New("Influx Error: Bucket required")
+	}
+
+	client := influxdb2.NewClient(i.metadata.URL, i.metadata.Token)
+	i.client = client
+	i.writeAPI = i.client.WriteAPIBlocking(i.metadata.Org, i.metadata.Bucket)
+	return nil
+}
+
+// GetInfluxMetadata returns new Influx metadata
+func (i *Influx) getInfluxMetadata(metadata bindings.Metadata) (*influxMetadata, error) {
+	b, err := json.Marshal(metadata.Properties)
+	if err != nil {
+		return nil, err
+	}
+
+	var iMetadata influxMetadata
+	err = json.Unmarshal(b, &iMetadata)
+	if err != nil {
+		return nil, err
+	}
+	return &iMetadata, nil
+}
+
+// Operations returns supported operations
+func (i *Influx) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{bindings.CreateOperation}
+}
+
+// Invoke called on supported operations
+func (i *Influx) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	var jsonPoint map[string]interface{}
+	err := json.Unmarshal(req.Data, &jsonPoint)
+	if err != nil {
+		return nil, errors.New("Influx Error: Cannot convert request data")
+	}
+
+	line := fmt.Sprintf("%s,%s %s", jsonPoint["measurement"], jsonPoint["tags"], jsonPoint["values"])
+
+	// write the point
+	err = i.writeAPI.WriteRecord(context.Background(), line)
+	if err != nil {
+		return nil, errors.New("Influx Error: Cannot write point")
+	}
+	i.client.Close()
```

Comment:
If this closes the channel, it means a second operation will fail, correct?

Reply:
Yes, indeed. I was not quite sure where the connection should be closed to be honest. Looking at other components such as MQTT, it's disconnecting in invoke as well. On second thought, probably should use defer here...

Reply:
Shouldn't close at all, otherwise the component will only ever work for one call to Invoke without restarting it.

Reply:
I tested the component and it did work well for a stream of data for several hours.

So I should just remove the close or is there a method to implement where I should close the connection?

Reply:
It looks like Close does not actually close the connection (which is just HTTP) but makes sure the data is written. This would explain why the subsequent calls succeed.

Reply:
Oh ok, that makes sense.